sklearn/tests/test_multiclass.py

```diff
@@ -921,3 +922,108 @@ def test_ovo_consistent_binary_classification():
     ovo.fit(X, y)
 
     assert_array_equal(clf.predict(X), ovo.predict(X))
+
+
+# Test that metadata is routed correctly for multiclass-classifiers
+# =================================================================
+
+
+@pytest.mark.usefixtures("enable_slep006")
+def test_fitted_coefs_differ_when_sample_weight_differs():
+    """Assert if coefficients differ when estimators are fitted with different
+    `sample_weights`."""
+    sample_weight = np.ones_like(iris.target, dtype=np.float64)
+    multiclass_classifiers = [
+        OneVsRestClassifier,
+        OneVsOneClassifier,
+        OutputCodeClassifier,
+    ]
+    for classifier in multiclass_classifiers:
+        clf = classifier(
+            estimator=SGDClassifier(random_state=42).set_fit_request(sample_weight=True)
+        )
+        # fitting with sample_weight=sample_weight
+        clf.fit(iris.data, iris.target, sample_weight=sample_weight)
+        # TODO: should we raise an error if sample weight is requested but is not
+        # passed in the router's fit method when enable slep006 is True?
+        # clf.fit(iris.data, iris.target)
+
+        # fitting with sample_weight=sample_weight_new
+        clf_with_different_weight = clone(clf)
+        sample_weight_new = sample_weight.copy()
+        sample_weight_new[::2] /= 2
+        clf_with_different_weight.fit(
+            iris.data, iris.target, sample_weight=sample_weight_new
+        )
+
+        # test if coefficients are different
+        for est, est_with_different_weight in zip(
+            clf.estimators_, clf_with_different_weight.estimators_
+        ):
+            # we need "if hasattr" because OutputCodeClassifier
+            # might make use of class _ConstantPredictor
+            if hasattr(est, "coef_") and hasattr(est_with_different_weight, "coef_"):
+                assert not np.allclose(est.coef_, est_with_different_weight.coef_)
+
+
+@pytest.mark.usefixtures("enable_slep006")
+def test_partial_fitted_coefs_differ_when_sample_weight_differs():
+    """Assert if coefficients differ when estimators are partial_fitted with different
+    `sample_weights`."""
+    sample_weight = np.ones_like(iris.target, dtype=np.float64)
+    multiclass_classifiers = [OneVsRestClassifier, OneVsOneClassifier]
+    for classifier in multiclass_classifiers:
+        clf = classifier(
+            estimator=SGDClassifier(random_state=42).set_partial_fit_request(
+                sample_weight=True
+            )
+        )
+        # partial_fitting with sample_weight=sample_weight
+        clf.partial_fit(
+            iris.data,
+            iris.target,
+            classes=np.unique(iris.target),
+            sample_weight=sample_weight,
+        )
```

Comment:
Migrating this test towards `sklearn/tests/test_metaestimators_metadata_routing.py` isn't easily done.

Reason: When partial_fitting `OneVsRestClassifier` and `OneVsOneClassifier` an additional param (`classes=np.unique(y)`) needs to be passed, that the central tests don't account for (btw. I wonder why the function isn't internally doing that): 

`sklearn/utils/multiclass.py:410: ValueError: classes must be passed on the first call to partial_fit.`

This error occurs when implementing the central test with `routing_methods`: `partial_fit` and the tests then call `OneVsRestClassifier.partial_fit() ` without `classes` (as they are supposed to).

In `test_setting_request_removes_error`, should we add a condition checking for additional params before calling `method(X, y, **kwargs)`? But then, in the next step, the methods of the artificial estimators from `sklearn/tests/metadata_routing_common.py `(for instance `ConsumingClassifier`) need to be changed as well, so they would accept `classes`. I don't know how many meta-estimator's methods take additional params like `classes`, but wouldn't it be quite a lot?

Reply:
this is a good one. We should certainly fix our common tests to handle cases when `partial_fit` accepts `classes` and when not. This could be done by a try/except catching `TypeError`. As for adding `classes` to other `partial_fit`, seems like a good idea, I wonder why they don't support it already.

Reply:
Okay, I will try that. 

But what is the reason why `partial_fit` allows the user to define `classes`?

I cannot think of any other way  than `classes=np.unique(y)`  to define classes. Why / in what case would a user want to deviate from that?

Reply:
So, I made something for `test_setting_request_on_sub_estimator_removes_error` that looks like this:

```
            try:
                method(X, y, **method_kwargs)
            except ValueError:
                method(X, y, classes=np.unique(y), **method_kwargs)
```

I get this error when I test this function.
`TypeError: ConsumingClassifier.partial_fit() got multiple values for argument 'sample_weight'`

I believe now I need to make the mocking estimators from `sklearn/tests/metadata_routing_common.py` (for instance `ConsumingClassifier`) to consume `classes` as well. This means I need to change all those that are classifiers and that are supposed to consume, because they could all be used for partial_fit. Can you confirm, @adrinjalali?

Also, I wonder if making the test and the mocking classifiers able to consume `classes` is general enough. If any metaestimator comes with their extra requirements, doesn't this get too crowded soon?

Reply:
So, I made something for `test_setting_request_on_sub_estimator_removes_error` that looks like this:

```
            try:
                method(X, y, **method_kwargs)
            except ValueError:
                method(X, y, classes=np.unique(y), **method_kwargs)
```

And `ConsumingClassifier` from `sklearn/tests/metadata_routing_common.py` now feeds `classes=None` in its `partial_fit` method.

It works at least. Please let me know if this is correct, @adrinjalali

Reply:
Cause the user might not have all the classes present in a batch of data passed to `partial_fit`. There can be a case where the batch being passed to `partial_fit` includes only a subset of total classes, and the classes are pre-defined by the user rather than selected from each batch.

Reply:
I had to do it like this, because `MultiOutputClassifier.partial_fit() ` expects its classes in another format.
```python
            try:
                method(X, y, classes=np.unique(y), **method_kwargs)
            except TypeError:
                method(X, y, **method_kwargs)
            except IndexError:
                method(X, y, **method_kwargs)
```